### ISSUE TITLE
Setting JAVA_HOME to the default Java changes functionality of Maven

### ISSUE THREAD

**Issue by datalogics-kam** [Homebrew/brew#226]
### Bug reports:

Something for your consideration/discussion:

``JAVA_HOME`` being unset is a useful state. The Homebrew glue script for Maven says:

```bash
#!/bin/bash
JAVA_HOME="${JAVA_HOME:-$(/usr/libexec/java_home)}" exec "/usr/local/Cellar/maven/3.3.9/libexec/bin/mvn" "$@"
```

This removes two layers of functionality:

- The ``mvn`` shell script already has code to figure out where Java is, including Darwin-specific code. But first it runs ``/etc/mavenrc`` and ``~/.mavenrc``
- My ``~/.mavenrc`` doesn't work the way it used to: I have mine set up to say something like "If this shell doesn't have ``JAVA_HOME`` set, and I'm on a Mac, use 1.7"

I know there's probably a _reason_ why Homebrew sets ``JAVA_HOME`` to the default, even if I don't understand it. It kind of confuses me, because if JAVA_HOME isn't set, then won't Java-based commands just use the default anyway?

At least for Maven, it seems redundant to put that in there. It also means that Maven-in-Homebrew doesn't function like just downloading Maven from Apache and putting it into the path, or like Maven does when I use the same dotfiles on, say, Linux.

I can work around it, so this isn't urgent or anything. But it does explain a mystery for me.

> We will consider your request but it may be closed if it's something we're not actively planning to work on.

As I said, I seek understanding, and this is only feedback. If it's really important that it work this way, I'll work around it.

Thanks for Homebrew; it's been quite useful to me.

**Comment by mikemcquaid** [comment]
CC @xu-cheng for more thoughts about why this changed happens; I'm personally lacking context here.

**Comment by apjanke** [comment]
TL;DR: What exactly do you want to accomplish here by leaving `$JAVA_HOME` undefined? I'll bet you can accomplish that through other means. Or if you can't, we can provide a purpose-designed hook for it. `brew` probably really does want to make sure that `$JAVA_HOME` is defined and controlled by `brew`, as part of setting up a uniform build environment.

Details follow:

------------

Homebrew takes responsibility for locating all the needed compilers and runtimes for software builds. it accepts user input, but doesn't just pass through the user's environment for build tools to use to do their own location logic. We try to reduce variance and impose control, and minimize sensitivity to user configuration.

It identifies and selects the JDK it's going to use for building, and setting `$JAVA_HOME` 

> because if JAVA_HOME isn't set, then won't Java-based commands just use the default anyway?

This begs the question of "what is the default, and how is it chosen?". And there's a lot of variance between Java packages' build processes, because Java never standardized any platform-specific deployment processes. (Much to its and our detriment, but it is what it is.)

As an experienced Java developer, I think Homebrew almost certainly wants to define and take responsibility for `$JAVA_HOME` within its builds. We can (and IMHO probably should) [respect user-supplied `$JAVA_HOME`](https://github.com/Homebrew/brew/pull/183) for location of JDK installations. But leaving `$JAVA_HOME` undefined and allowing Java packages to do their own JDK detection is a recipe for unending pain, for Homebrew users and maintainers both.

Homebrew definitely should not be sensitive to a user-specific `~/.mavenrc`. I would also argue that it should not be sensitive to a machine-specific `/etc/mavenrc`, either. In Maven terms, Homebrew produces globally-relevant build artifacts that are relocatable to all users. `~/.mavenrc` and `/etc/mavenrc` are relevant for developers. If they provide caching and upstream provider configuration, we should respect that, but that's mostly for performance. Anything that changes the actual outcome of a build is something Homebrew would want to filter out. This is analagous to changing the location of the upstream `homebrew/brew` repo and bottle domain: probably useful to some people, but an advanced configuration, and something we should reject hard in the normal use case.

So, _all that being said_: yeah, Homebrew's interaction with Maven is a little half-assed. But Homebrew's goal is to minimize user-dependent variability, and to provide standard quasi-system-level builds of stuff. I think `$JAVA_HOME` JDK detection and Maven configuration are largely orthogonal: Maven is mostly concerned about artifact identification and caching, and has a very good system for that. If we can align our sandboxing with the Maven repo caching configuration, that could make our builds faster and smoother. But in general, Maven configuration that alters the outcome of a build, as opposed to how you fetch centrally-tracked artifacts, is non-brewy.

**Comment by apjanke** [comment]
Put another way: When you are installing Java software using Homebrew, `brew` is the primary installation tool, and its use of Maven is an internal implementation detail, which should not be sensitive to the user environment except in ways that `brew` provides as part of its public API. Homebrew installs packages for users, not developers of those packages.

IMHO, the whole fact that a package is even implemented in Java should be basically transparent to end users.

**Comment by UniqMartin** [comment]
I believe @apjanke has provided an extensive answer. Feel free to follow up, if there's still something not addressed by it. But I'm closing this now, as I don't see something actionable (and no bug awaiting a fix).
